pkg/syndesis/action/upgradebackoff.go

```diff
@@ -0,0 +1,66 @@
+package action
+
+import (
+	"github.com/operator-framework/operator-sdk/pkg/sdk"
+	"github.com/sirupsen/logrus"
+	"github.com/syndesisio/syndesis-operator/pkg/apis/syndesis/v1alpha1"
+	"math"
+	"time"
+)
+
+const (
+	BackoffCap = 8 * time.Minute
+)
+
+// After a failure, waits a exponential amount of time, then retries
+type UpgradeBackoff struct {
+	operatorVersion	string
+}
+
+func (a *UpgradeBackoff) CanExecute(syndesis *v1alpha1.Syndesis) bool {
+	return syndesisInstallationStatusIs(syndesis, v1alpha1.SyndesisInstallationStatusUpgradeFailureBackoff)
+}
+
+func (a *UpgradeBackoff) Execute(syndesis *v1alpha1.Syndesis) error {
+
+	now := time.Now()
+
+	lastFailureWrapper := syndesis.Status.LastUpgradeFailure
+	var lastFailure time.Time
+	if lastFailureWrapper != nil {
+		lastFailure = lastFailureWrapper.Time
+	} else {
+		lastFailure = time.Now().Add(- 8 * time.Hour)
+	}
+
+	if lastFailure.After(now) {
+		lastFailure = now
+	}
+
+	power := float64(syndesis.Status.UpgradeAttempts - 1)
+	if power < 0 {
+		power = 0
+	}
+
+	delay := time.Duration(math.Pow(2, power)) * time.Minute
+	if delay > BackoffCap {
```

Comment:
Shouldn't we maybe at some point just stop trying to upgrade if a certain amount of retries has happened ? and then enter a final state like 'CrashLoopback' where no retry is happening again ...

Reply:
Yes, I was unsure about it, since it would have been a state with no possible recover.. But since upgrading means scale down then up, I think doing it over and over again is probably worse.. so I'll add it..